src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,407 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/relax/transform/gradient.cc
+ * \brief A reverse-mode automatic differentiation.
+ *
+ * Now only supports differentiating a function in the IRModule with one dataflow block
+ * with respect to the only return value of the function. It needs to be scalar.
+ *
+ * For examples, see the MLP examples in tests/python/relax/test_transform_gradient.py and
+ * tests/python/relax/test_transform_gradient_numeric.py.
+ */
+
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/op_attr_types.h>
+#include <tvm/relax/transform.h>
+
+#include <unordered_set>
+
+namespace tvm {
+namespace relax {
+
+class GradientMutator : public ExprMutator {
+ public:
+  explicit GradientMutator(IRModule mod, const Array<Var>& require_grads)
+      : ExprMutator(mod), require_grads_(require_grads) {}
+
+  Function FuncTransform(const FunctionNode* node) {
+    ICHECK(node->body->IsInstance<SeqExprNode>());
+
+    Array<Var> new_params;
+    for (Var param : node->params) {
+      Var new_param = Var(param->vid, param->shape(), param->checked_type(), param->span);
+      this->var_remap_[param->vid] = new_param;
+      new_params.push_back(new_param);
+    }
+    Expr new_body = VisitWithNewScope(node->body);
+
+    const SeqExprNode* seq_expr = new_body.as<SeqExprNode>();
+    // only support a single dataflow block
+    ICHECK(seq_expr->blocks.size() == 1);
+    ICHECK(seq_expr->blocks[0]->IsInstance<DataflowBlockNode>());
```

Comment:
It is better to do such check before mutation. So you might want to use the pattern
```c++
CHECK(cond1) << "error msg";
CHECK(cond2) << "error msg";
...
mutate();
```

ICHECK means “internal CHECK”. Here we should use CHECK instead, because we require the user input to meet the requirement.

Reply:
Wait. What’s the purpose of the recursive mutation on line 54? I don’t see any override of `VisitExpr_` so supposably this recursion changes nothing.

Reply:
Do you mean the `VisitWithNewScope`? This is because I set `var_remap_` and want to replace the variables with new ones

Reply:
Check before mutation is a good idea but it is convenient to do checks immediately to get `body`, `seq_expr` and `block` when mutating. If I check before mutation, I should go through this again.

Reply:
I see. Do you think it is a good idea to move this part of logic to a stand-alone helper function, likely, named “CopyFuncWithFreshParam”? So that we don’t mix all logic into a single function.

Reply:
And I believe these checks are temporary because now we can only do single-dataflowblock AD. So shoud I change it to checking before mutation pattern?

Reply:
It seems that so far the only purpose of having the ExprMutator is for this parameter replacement. I would suggest having the standalone function for copy and simplifying the logic of the gradient function generation.

Reply:
> I see. Do you think it is a good idea to move this part of logic to a stand-alone helper function, likely, named “CopyFuncWithFreshParam”? So that we don’t mix all logic into a single function.

Looks good

Reply:
Wait a minute. I see you emitting new bindings after the original binding. In this case, let’s keep it as it is now, since moving the copy to a standalone function means we will do a recursive mutation there, and afterwards we will do another recursive mutation here.

Reply:
Let me remove the helper function then